synapse/lib/lmdbslab.py

```diff
@@ -166,6 +166,135 @@ def inc(self, name, valu=1):
         self.set(name, curv)
         return curv
 
+class SafeKeyVal:
+    '''
+    Key/value storage that does not store items in memory and safely handles large keys.
+    '''
+    def __init__(self, slab, name, prefix=None):
+
+        self.name = name
+        self.slab = slab
+        self.valudb = slab.initdb(name)
+
+        self.prefix = prefix
+        if prefix is not None:
+            self.preflen = len(prefix)
+
+    def getSubKeyVal(self, prefix):
+
+        if self.prefix is not None:
+            prefix = self.prefix + prefix
+
+        return SafeKeyVal(self.slab, self.name, prefix=prefix)
+
+    def get(self, name, defv=None):
+        '''
+        Get the value for a key.
+
+        Note:
+            This may only be used for keys < 512 characters in length.
+        '''
+        if self.prefix is not None:
+            name = self.prefix + name
+
+        if len(name) > 511:
+            raise s_exc.BadArg('SafeKeyVal.get() may only be used with keys < 512 characters in length.')
```

Comment:
These BadArg values should include the very long name. That could probably be truncated at 1024 chars.